### PR TITLE
Added min size for file systems

### PR DESCRIPTION
**Pull request by japokorn**
- added default minimal file system size into FS class (2 MiB)
- overriden in certain subclasses (e.g. NoDevFS)
- updated tests behavior to reflect the changes
    
Resolves: #1554519

### CODE REVIEW

File: blivet/formats/fs.py [hunk 1]
```diff
@@ -830,6 +831,7 @@ class Ext2FS(FS):
     _check = True
     _fsck_class = fsck.Ext2FSCK
     _info_class = fsinfo.Ext2FSInfo
+    _min_size = Size("2 MiB")
```
Comment (dwlehman) [inline blivet/formats/fs.py]: This one is unneeded because of the one in `FS` above.

Comment (japokorn) [reply blivet/formats/fs.py]: Fixed.

File: blivet/devices/lvm.py [hunk 2]
```diff
@@ -1901,7 +1901,7 @@ def _set_size(self, newsize):
     @type_specific
     def max_size(self):
         """ The maximum size this lv can be. """
-        max_lv = self.size + self.vg.free_space
+        max_lv = self.vg.align(self.size, roundup=True) + self.vg.align(self.vg.free_space)
```
Comment (vojtechtrefny) [inline blivet/devices/lvm.py]: This definitely needs some explanation -- how changing filesystem minimal size influences max size of an LV?

File: tests/devicetree_test.py [hunk 3]
```diff
@@ -32,7 +33,7 @@ def test_resolve_device(self):
         dev1_label = "dev1_label"
         dev1_uuid = "1234-56-7890"
         fmt1 = get_format("ext4", label=dev1_label, uuid=dev1_uuid)
-        dev1 = StorageDevice("dev1", exists=True, fmt=fmt1)
+        dev1 = StorageDevice("dev1", exists=True, fmt=fmt1, size=Size("2 MiB"))
```
Comment (vojtechtrefny) [inline tests/devicetree_test.py]: It would be better to use `fmt1.min_size` here in case we decide to change it in the future.

File: blivet/devices/lvm.py (hunk 2 above)
Comment (japokorn) [reply blivet/devices/lvm.py]: Probably some rogue line of code. Removed.

File: tests/devicetree_test.py (hunk 3 above)
Comment (japokorn) [reply tests/devicetree_test.py]: Fixed.

File: blivet/devices/lvm.py (hunk 2 above)
Comment (dwlehman) [reply blivet/devices/lvm.py]: The changes to the tests seem to have revealed a bug in the lv max_size calculation. If the lv size is unaligned, you end up with a bit of extra space available that isn't accounted for in max_size. In the test I was doing, the pv was 1 GiB and the lv was 499MiB. The vg should have had 520GiB free (1GiB - 1 extent - lv size), which it did. It was reporting a max_size of 1019MiB (lv size + vg free). However, there is the remaining 1MiB in the lv's last extent that should be taken into account as well, making the max_size for the lv 1020MiB. The current/upstream code was reporting it as 1019 instead. This should be split out to a separate commit with a good explanation in the commit message.

Comment (japokorn) [reply blivet/devices/lvm.py]: I have split PR to two commits. The second one describes the uncovered issue and fixes it.

File: blivet/devices/lvm.py [hunk 4]
```diff
@@ -1951,7 +1951,7 @@ def _set_size(self, newsize):
...
```
[the remark this answers is not in the document]
Comment (japokorn) [reply blivet/devices/lvm.py]: Fixed. Rewritten the commit message completely.
Added `roundup=False` for clarity.

File: blivet/blivet.py [hunk 5]
```diff
@@ -1190,17 +1190,7 @@ def factory_device(self, device_type=devicefactory.DEVICE_TYPE_LVM, **kwargs):
 
     def copy(self):
         log.debug("starting Blivet copy")
-
```
Comment (vojtechtrefny) [inline blivet/blivet.py]: This change looks unrelated, probably some rebase related issue.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

File: tests/formats_test/fstesting.py [hunk 6]
```diff
@@ -87,9 +87,6 @@ def test_instantiation(self):
         self.assertEqual(an_fs.resizable, False)
 
         # sizes
-        expected_min_size = Size(0) if can_resize(an_fs) else an_fs._min_size
-        self.assertEqual(an_fs.min_size, expected_min_size)
```
Comment (dwlehman) [inline tests/formats_test/fstesting.py]: Should we not continue to verify this? The first expression would be simplified to `expected_min_size = an_fs._min_size` and the second line would remain as it is.

File: tests/devices_test/lvm_test.py [hunk 7]
```diff
@@ -87,13 +87,16 @@ def test_lvmcached_logical_volume_init(self):
         vg = LVMVolumeGroupDevice("testvg", parents=[pv, pv2])
 
         cache_req = LVMCacheRequest(Size("512 MiB"), [pv2], "writethrough")
-        lv = LVMLogicalVolumeDevice("testlv", parents=[vg],
+        lv = LVMLogicalVolumeDevice("testlv",
+                                    parents=[vg],
                                     fmt=blivet.formats.get_format("xfs"),
-                                    exists=False, cache_request=cache_req)
+                                    size=Size("2 MiB"),
```
Comment (dwlehman) [inline tests/devices_test/lvm_test.py]: I think it would be better if the size here was expressed as `pv.format.min_size` or similar for clarity.

File: tests/devices_test/device_properties_test.py [hunk 8]
```diff
@@ -759,6 +762,8 @@ def __init__(self, methodName="run_test"):
             "parents": xform(lambda x, m: self.assertEqual(len(x), 1, m) and
                              self.assertIsInstance(x, ParentList) and
                              self.assertIsInstance(x[0], LVMVolumeGroupDevice)),
+            "size": xform(lambda x, m: self.assertEqual(x, Size("2 MiB"), m)),
+            "target_size": xform(lambda x, m: self.assertEqual(x, Size("2 MiB"), m))
```
Comment (dwlehman) [inline tests/devices_test/device_properties_test.py]: Would be better if instead of `Size("2 MiB")` you used `get_device_format_class("fs")._min_size` or similar.

**Comment by dwlehman** [comment]
Ugh, apparently github shouldn't be offering per-commit reviews since it will trample earlier commit reviews with the latest one.

Please take a look at my comments on the first/main commit.

**Comment by japokorn** [comment]
jenkins, test this please

File: tests/devices_test/lvm_test.py (hunk 7 above)
Comment (japokorn) [reply tests/devices_test/lvm_test.py]: Fixed.

File: tests/formats_test/fstesting.py (hunk 6 above)
Comment (japokorn) [reply tests/formats_test/fstesting.py]: Fixed.

File: tests/devices_test/device_properties_test.py (hunk 8 above)
Comment (japokorn) [reply tests/devices_test/device_properties_test.py]: Fixed.

File: tests/devices_test/device_properties_test.py [hunk 9]
```diff
@@ -79,10 +80,8 @@ def __init__(self, methodName='run_test'):
             "path": xform(lambda x, m: six.assertRegex(self, x, "^/dev", m)),
             "raw_device": xform(self.assertIsNotNone),
             "resizable": xform(self.assertFalse),
-            "size": xform(lambda x, m: self.assertEqual(x, Size(0), m)),
             "status": xform(self.assertFalse),
             "sysfs_path": xform(lambda x, m: self.assertEqual(x, "", m)),
-            "target_size": xform(lambda x, m: self.assertEqual(x, Size(0), m)),
```
Comment (dwlehman) [inline tests/devices_test/device_properties_test.py]: Why did you have to remove these from the base class and add them to the various subclasses?

Comment (japokorn) [reply tests/devices_test/device_properties_test.py]: Fixed: Moved the lines back to base class. Only those with non-default values were left in subclasses.
